### DISCUSSION TITLE
Eye candy (candies)

### DISCUSSION THREAD

**Discussion by udippel** [gerbera/gerbera#2867] [General]
(Before boredom sets in ...)

The Database view contains a number of entries looking like this:
![image](https://github.com/gerbera/gerbera/assets/50265752/998d56a0-3d6d-482f-b9e7-d1f91aaacc8d)
The underlying code can be seen here:
![image](https://github.com/gerbera/gerbera/assets/50265752/a7706bdc-e4de-4cd8-9e9e-892c3a2ea94f)
There seems to be no such directory/file in the file system.

Actually, why are some entries read out from the database with icons other than the folder icon in front of these? I guess, these are titles that have some icons associated with them in the database. Though, as far as I am concerned, they do not server much purpose in this display?

**Comment by KarlStraussberger** [comment]
I've got a simple answer: The media file says it contains an image (aka cover art). But that once downloaded it cannot be displayed. So you can check out the file in filesystem browser (e.g. Explorer) in look at the details or use something like foobar2000 to embed further/correct images.

**Reply by udippel** [reply]
So I guess. Still, to me they seem to serve no purpose here. Some actually do display. Some show the 'missing icon', but eventual do show proper icons in front of the individual files in the same Database view.
With 10.000+ files, and in this view, I'd prefer to have a switch (do we have one?) to just turn this off instead of browsing around to find out which images display, which exist and in which format. I was never in album art. 
A switch would be nice, or a trap that skips the icon display in case of 'no icon'? What do you think?

**Reply by KarlStraussberger** [reply]
Actually, there is a setting for your case: https://docs.gerbera.io/en/latest/config-server.html#ui `show-thumbnail`

**Reply by udippel** [reply]
Fine! Thanks, slowly learning the documentation .... So we DO have this switch.

**Reply by udippel** [reply]
Now I need help: 
What's wrong with 
`<show-thumbnail>enabled="no"></show-thumbnail>
`
in ui? They still show.

**Reply by KarlStraussberger** [reply]
Full example
```
<ui enabled="yes" show-numbering="yes" show-tooltips="yes" poll-when-idle="no" poll-interval="2" show-thumbnail="no">
  <accounts enabled="no" session-timeout="30">
    <account user="mediatomb" password="mediatomb"/>
  </accounts>
  <items-per-page default="50">
    <option>10</option>
    <option>25</option>
    <option>50</option>
    <option>100</option>
  </items-per-page>
</ui>
```

**Comment by udippel** [comment]
Oh, thanks a lot! Hadn't understood that.

**Comment by udippel** [comment]
Okay, one more. Minor, 'eye candy',and yet (in MHHO) an inconsistency. In Config, I can in UI untick the "Show Track Numbering". There is no "Apply", so I can only go back - and still see the track numbering. And when returning to Config, the "Show Track Numbering" is still ticked. 
If we can't change it from here, and can't 'apply', then, by all means, let's not show this as clickable; but as grayed out.

**Reply by KarlStraussberger** [reply]
There is a "save config". There are some config items that even need a restart of the server - which is not possible/visible via ui . In that specific case you at least have to force reload your browser to get rid of cached data.

**Reply by udippel** [reply]
Never mind. Then let's make them red, and add the footnote that these require a server restart. Problem solved. Nobody would know otherwise.

**Comment by udippel** [comment]
Please accept my apologies here, I had overlooked that Save Config. My mistake. It does work.

**Comment by udippel** [comment]
Saw a new update to 60. Any specifics; change log?

**Reply by KarlStraussberger** [reply]
Just made it 61. 
- Generation of example configuration
- Removing entries not found on disk
- Update of entries in Config UI

**Reply by udippel** [reply]
Thanks. Seems to be much more friendly, when I uninstall and reinstall with each iteration.

1. Question for proper handling: Once I remove an item from Database, what's the proper way to bring it back? Add Item or Add Autoscan Item? Seems unclear, since the item falls into a directory whose root is autoscan. Does it inherit the autoscan of its parent at re-appearance?
2. Likewise, when Recently Modified is removed, how to bring it back?
3. Expiration time of Recently ... is mentioned in the documentation, but I don't seem to be able to find this in Config. I think that's a config the average user might want to play with? Also the number of items.
4. The working of Timed: When is  (e.g.) '1800 seconds'? When is 0 seconds? In other words: At least personally I'd prefer a crontab-like setting, when the user can state the time of autoscan. For example - also for resource reasons - at 04:00. 
5. Where do I find the example configuration?

Looks like good progress!

**Reply by KarlStraussberger** [reply]
> Thanks. Seems to be much more friendly, when I uninstall and reinstall with each iteration.

No need for that. Only in rare cases we remove files. Normally, apt should do that by itself

> 1. Question for proper handling: Once I remove an item from Database, what's the proper way to bring it back? Add Item or Add Autoscan Item? Seems unclear, since the item falls into a directory whose root is autoscan. Does it inherit the autoscan of its parent at re-appearance?

If you remove from "pc directory" (physical view) it will appear with the next startup.
If you remove from a virtual folder it will only reappear if you touch the file or rename it and thus trigger the virtual layout logic.

> 2. Likewise, when Recently Modified is removed, how to bring it back?

It apprears after restart -> should not be removable in the first place!

> 3. Expiration time of Recently ... is mentioned in the documentation, but I don't seem to be able to find this in Config. I think that's a config the average user might want to play with? Also the number of items.

Its all here and even contains an example
https://docs.gerbera.io/en/latest/config-server.html#containers

> 4. The working of Timed: When is  (e.g.) '1800 seconds'? When is 0 seconds? In other words: At least personally I'd prefer a crontab-like setting, when the user can state the time of autoscan. For example - also for resource reasons - at 04:00. 

On my private requirement list for a long time. But time required is what's lacking...

> 5. Where do I find the example configuration?

run `gerbera --create-example-config`

> Looks like good progress!

**Reply by udippel** [reply]
Please read carefully. I wrote that the "Recent ..." stuff was in the documentation, I only had not seen it in the web-based Config. To me, this is really something that should be there for users to adjust according to their preferences. It is by far the easiest way to see if new items have been picked or, respectively deleted (those don't show as such, AFAICS). Could be really informative, if adjustable. The most recent 20 or 50, sorted according to time, might be fast and useful. Instead of having to open a client, e.g.

**Reply by udippel** [reply]
So far 61 here does not remove items not found on disk. Additions work fine, immediately, removal not. Do I have to wait longer? What for, since I don't run Timed here. 
[Edited:] I am aware that Inotify isn't most logical on a read-only file system. Since, however, against my expectations the newly added files while gerbera is off, and the drive off-site for update, ARE picked up at restart: are there different rounds of grb for 'adding' and 'removal': or should grb pick up both kind of changes once it runs? 
If so, I'd change to Timed, for testing purposes.

**Reply by udippel** [reply]
May I refer to my earlier question, please? Adding an item: does it inherit the autoscan properties of its parent?

**Reply by KarlStraussberger** [reply]
If you have autoscan enabled you don't have to add manually at all.

In fact, inotify and timed share the same startup. During startup new items are added and deleted ones should be identified to be deleted at the end of the initial scan. That means that the situation needs further investigation.

**Reply by udippel** [reply]
"If you have autoscan enabled you don't have to add manually at all." That's obvious.

I was asking for that further investigation. I plan to:
1. Delete the sample directory from the Database view.
2. Check if it was actually gone (eventually with restart of gerbera?)
3. Go to File System, and add that directory once again. Exactly here is the reason for my question: Add Item? or Add Autoscan Item? I think this is a most relevant question here: "To inherit or not to inherit" is relevant.
4. Check if it is okay now.

Definitely these steps must necessarily work; read-only file system or not. 
Before doing it, just wanted to make sure how I should re-add that sample directory with modified contents.

**Reply by KarlStraussberger** [reply]
> "If you have autoscan enabled you don't have to add manually at all." That's obvious.
> 
> I was asking for that further investigation. I plan to:
> 
>     1. Delete the sample directory from the Database view.

You have to delete it from "PC Directory".

>     2. Check if it was actually gone (eventually with restart of gerbera?)
 
If you don't see errors on log, it's gone

>     3. Go to File System, and add that directory once again. Exactly here is the reason for my question: Add Item? or Add Autoscan Item? I think this is a most relevant question here: "To inherit or not to inherit" is relevant.

If you restart, it will be detected automatically. If you "Add Item" no change detection is associated, so you have to "Add Autoscan" (if not covered by config-entry).

>     4. Check if it is okay now.
> 
> 
> Definitely these steps must necessarily work; read-only file system or not. Before doing it, just wanted to make sure how I should re-add that sample directory with modified contents.

**Reply by udippel** [reply]
I was under the impression, the method was to delete from the e.g. virtual folders, and these being the "real" database folders.
Thanks for the clarification! (I hope I'm not the first and only one to misread this aspect.)

Now I am even more confused about the "Add Item" and "Add Autoscan Item"; though that's not the topic in this moment.

**Reply by udippel** [reply]
I will happily confirm that the method as indicated above did work!

gerbera spew a lot of errors during the time BEFORE, on trying to access those deleted files with BubbleUPnP, like this one:
`023-09-01 08:16:16  error: [/build/src/file_request_handler.cc:105] virtual void FileRequestHandler::getInfo(const char*, UpnpFileInfo*) Error: Failed to open /.../RP_2017_4_Italies_2.mp3: No such file or directory`
as could be expected.
Now the log is clean, and the display in the web ui as well as on BubbleUPnP reflects the contents of the directories correctly.

**Comment by udippel** [comment]
Still thinking ...
So 'Delete' does not mean this request will be stored and kept beyond the next restart? Okay, but that'd be anything but obvious. 
If Images is an autoscan directory, I could not eternally exclude the images of my wife's presents before her birthday; as long it is a directory under Images? 
That's why I was so obsessed with "How to get it back?" - It does come automatically! Never expected that.
Though, what's the point of Delete then? It would be a 'Delete for this session'; correct? I'd say 'no purpose', at least at first sight. 
And this would also answer my earlier question: Yes, Autoscan is inherited by all subdirectories that fall under autoscan. Which does make a lot of sense.

**Reply by KarlStraussberger** [reply]
That's the reason for virtual layout. You can hide PC directory from browsing, so deleting virtual entries has the desired effect.

The point is: How can gerbera know whether an entry was deleted in the last session or added since the last run. There is no such thing as remember the deleted entries forever.

**Reply by udippel** [reply]
Okay, the item with the virtual folders makes lot of sense. Though it is anything but obvious. The casual user would have to kneel very deep into this. I'll have to think further ... .

With respect to 'remember': I had not expected any different. Though there must be a method, or a routine that can be called (by crontab ;-) ?) to compare, to verify the contents. I agree, that's different from Add, or adding items. Though it ought to exist. 
Can you confirm that it does work when a read-write device is connected, and an item is deleted when gerbera is 'live' through a file utility? If so, it would be okay for the moment. 
Though a verification should remain on the wish list, because it can happen in any read-write session, too, that a luser just pulls out the drive, does whatever, and plugs it back. Would never be synchronous again.
And the command-line based rm gerbera.db* isn't for everyone. So some 'start from scratch' ought to be available, I think. With the - preferred - alternative of a regular or user-initiated verification. 
gerbera does a great job in this respect already, if one considers that my thin client matchbox can pick up > 2 TB of data containing > 12.000 files from scratch within about 1 hour. Then, verification of just the further existence(!) of some item should be possible within a minute or two. 

[Edited:] This brings me back to my previous wishlist item of a button for "Update database". Here any user could verify database against actual drive contents, guaranteeing that both are in synch at the moment of his or her desire.

**Reply by KarlStraussberger** [reply]
From a long time user's perspective I cannot share your experience: When files are removed from the disk they are deleted from the database: Either Inotify detects this or it happens on startup. There may be one or the other issue in the new import mode but in general it's working as expected. "Update database" in what happens all the time as long as you do not mess with the database manually.

**Reply by udippel** [reply]
Tried twice, and twice removed files were not detected as such on startup. Just as new files were detected with reliability. No 'messing' here. Just started from scratch, except of what we did further up. 
If it did a scan for existence at startup, it should be okay. Seemingly this didn't happen here, twice (since I can only add and remove when gerbera is down).

**Comment by udippel** [comment]
"error: Object not found: 11740" in the log means what, please? I get a good dozen of these, without actually encountering any missing item or other error on the client side that I´d be aware of? How can I find out which item that is; if it is a specific media item after all?

**Reply by KarlStraussberger** [reply]
That happens after items being deleted are still referenced in virtual layout. I know there are some small bugs depending on the sequence items are deleted/disappearing. Up to now we only saw the effect but could not work out why it happens.

**Reply by udippel** [reply]
Second half of the question, please: is there any sql command to find out which item this refers to? I'd like to find out if it still exists. I did a restart in the meantime, though the same thing keeps happening.
Again, here I'd prefer a 'Update Database' instead of 'rm gerbera.db*', which would probably sort this item out, but also removed some tweaks and hides. 
At least, the restart of gerbera on a read-only file system did not notice a missing item.

**Reply by KarlStraussberger** [reply]
The SQL statement is not that simple. It's in table `grb_cds_resource` column `fanArtObject`. The `item_id` points to `mt_cds_object` which either is the phyiscal item (location starts with `F`) or the virtual item (location starts with `V`). In the virtual case the link to the physical item is in column `mt_cds_object.parent_id`

**Reply by udippel** [reply]
Played a bit with this, but ended in Nirwana. Never mind.
